Issue #722 Terracotta clustering service configuration support

cljohnso opened the pull request:
This PR is in two parts:

1. Refactors EhcacheManager to split CachePersistenceManager methods out.  This creates a BaseCacheManager abstract class holding the non-persistent methods leaving EhcacheManager as a subclass maintaining the current functionality. This is a positioning move for supporting clustering.

  * Adds requiresSerialization method to ResourceType
  * Splits EhcacheManager.newStore into BaseCacheManager (non-DISK)
  * Generalizes CacheManagerBuilder to
    * return builder-designated CacheManager type (subclass)
    * return CacheManagerConfiguration-driven CacheManagerBuilder subclass
    * adds newSpecializedCacheManager method to obtain type-specific CacheManager from configuration

1. Adds interfaces and concrete classes in support of configuring and
building a clustered CacheManager.

  * ehcache-clustered-ext.xsd defining the <service> element extension
  * ClusteringServiceConfigurationParser for parsing XML <service> extension
  * ClusteringServiceConfiguration providing both ServiceCreationConfiguration and CacheManagerConfiguration support
  * ClusteringService, DefaultClusteringService and ClusteringServiceFactory providing the anchor for the clustering service configured with the ClusteringServiceConfiguration
  * ClusteredCacheManager, ClusteredCacheManagerImpl, and ClusteredCacheManagerBuilder providing support for a clustered CacheManager

cljohnso commented:
The _max memory_ commit is expected to be replaced by something more finely-tuned.

lorban commented on clustered/client/src/main/java/org/ehcache/clustered/ClusteredCacheManagerImpl.java:
```diff
@@ -0,0 +1,81 @@
+
+  @Override
+  public Maintainable toMaintenance() {
+    return null;
```
I'd rather throw UnsupportedOperationException for now.

lorban commented on clustered/client/src/main/java/org/ehcache/clustered/ClusteredCacheManagerImpl.java:
```diff
@@ -0,0 +1,81 @@
+  }
+
+  @Override
+  public void destroyCache(final String alias) throws CacheDurabilityException {
```
Same as above.

cljohnso replied:
`destroyCache` has nothing to do at this point .. should it really throw an `UnsupportedOperationException` now?

chrisdennis replied:
I'm happy with either... I imagine this going to see some significant refactoring when you and Louis start looking at the tighter integrations of store factories with resource and the various persistence mechanisms.

cljohnso replied:
I think I'll leave `destroyCache` as-is (no action) for the time being -- there's presently nothing to destroy so the empty method isn't, technically, incorrect.  `toMaintenance`, on the other hand, will throw.

lorban commented on clustered/client/src/main/java/org/ehcache/clustered/config/ClusteringServiceConfiguration.java:
```diff
@@ -0,0 +1,68 @@
+ *
+ * @author Clifford W. Johnson
+ */
+// TODO: Should this accept/hold a *list* of URIs?
```
This should be a Set of URIs IMHO.

cljohnso replied:
For a `Set`,  `ClusteringServiceConfigurationParser#parseServiceCreationConfiguration` will need to do a check for duplicates ... or not.  Should we fail a configuration with duplicate entries or just accept what fits in the set, first come/first served.  Should there be any attempt at resolving hosts during validation?  Should the set constructed from an XML configuration be _ordered_ (e.g. a `LinkedHashSet`) so the iteration order over the set will reflect the ordering in the XML?

chrisdennis replied:
I would stick with one URL here for the moment.  We know that AA and AP might change this, but for simplicity I'd stick with one until forced to change.

cljohnso replied:
We can certainly do that and, when the time comes, handle the `Set` version as an extension to this class.  The extension schema (as it currently stands) will need to be changed to increase the `maxOccurs` for the `connection` element -- also upward compatible.

cljohnso replied:
On the other hand, should I change the field holding the connection URI to a `Set` or `List` and changing the associated getter?  I believe the ultimate target is to consume more than one connection URL (for AA & AP) -- have we any concern/desire to position the consumers of this bit of configuration for that eventuality?

lorban commented on clustered/client/src/main/java/org/ehcache/clustered/config/xml/ClusteringServiceConfigurationParser.java:
```diff
@@ -0,0 +1,104 @@
+            try {
+              connectionUri = URI.create(urlValue);
+            } catch (IllegalArgumentException e) {
+              throw new RuntimeException(
```
Throwing RuntimeException never is a brilliant idea especially since it is also done in other locations and there's no specialized exception. IMHO this is an indicator that a little bit of refactoring should happen here, but outside the context of this task.

cljohnso replied:
This was modeled after code in `org.ehcache.transactions.xml.TxCacheManagerServiceConfigurationParser` -- which throws `RuntimeException`.  Should we settle on something else?

chrisdennis replied:
I don't think a RuntimeException is terrible here since this is startup time code... but maybe we should standardize on something more specific?  @ljacomet, thoughts?

lorban replied:
FIY I was thinking about something simple like XMLConfigurationException that extends from RuntimeException.

cljohnso replied:
I've added an `XmlConfigurationException` to the **xml** module ... looks like there are a number of places where it might be brought to bear.  Should I create an issue to cover that work?

chrisdennis replied:
Yes.

lorban commented on core/src/main/java/org/ehcache/core/BaseCacheManager.java:
```diff
@@ -0,0 +1,634 @@
+  private final ClassLoader cacheManagerClassLoader;
+  private final boolean useLoaderInAtomics;
+
+  private final Logger logger = LoggerFactory.getLogger(this.getClass());
```
I think BaseCacheManager.class would be better than getClass()

cljohnso replied:
A discussion with @chrisdennis led to this.  I'll follow any agreed upon model.

ljacomet replied:
If I understand the change here, the goal is to have the logging appear as coming from the sub type. Not sure I like it that much. Unless we decide that logs should be "functional" in nature, that is abstract the code getting executed ... but that seems weird.

cljohnso replied:
@chrisdennis ... would you like to weigh in on this?

I came from a shop where we didn't want to expose structural/compositional details in our logging ... we went as far as defining loggers at the **interface** level -- logical vs "physical" logging, if you will.  Like I said, I'll follow any agreed upon model.

chrisdennis replied:
To my mind using the concrete type for the logger makes sense... but it does make the logging code itself more convoluted.  For that reason maybe either a private logger per type in hierarchy makes more sense... but either way we should be consistent.

cljohnso replied:
For my two cents, I think using `BaseCacheManager` (the "parent" class name) hides information potentially useful in problem source identification -- what we really need to know is the _actual_ class responsible for getting is here (`EhcacheManager` or `ClusteredCacheManager` in this case) ... simply using `BaseCacheManager` hides that so we have to rely on clues obtained from elsewhere to know which of the cache manager implementations is involved.

lorban commented on core/src/main/java/org/ehcache/core/EhcacheManager.java:
```diff
@@ -111,76 +61,26 @@ public EhcacheManager(Configuration config) {
         destroyPersistenceSpace(alias);
       } catch (CachePersistenceException e) {
-        LOGGER.debug("Unable to clear persistence space for cache {}", alias, e);
+        this.getLogger().debug("Unable to clear persistence space for cache {}", alias, e);
```
Why this.getLogger()? I also wonder if this shouldn't be a warning instead of a debug.

cljohnso replied:
The use of `debug` was pre-existing.  As for `Logger` access, I'll follow any agreed upon model.

ljacomet replied:
It is a debug because there is not a lot to do here. If the file cannot be destroyed, it will be an issue only if you try re-creating in the same location.

lorban replied:
Regarding debug vs warn, I generally am in favor of warning the user when something not fatal but worth looking into happened like here.

cljohnso replied:
I agree with @lorban on this one ... something non-fatal happened that might deserve attention by the user.  We're logging at an info level cache creation, cache closure, and (normal) cache removal -- so this can't be a concern about noise.

This is pre-existing code but I'm going to change this to warn unless someone has a strong objection.

lorban commented on core/src/main/java/org/ehcache/spi/cache/ObservableCacheManager.java:
```diff
@@ -0,0 +1,31 @@
+ *
+ * @author Clifford W. Johnson
+ */
+public interface ObservableCacheManager {
```
This should extend CacheManager IMHO.

cljohnso replied:
Anyone else want to weight in on this?  @chrisdennis @ljacomet ?  I have no objection to making this extend `CacheManager` ...

chrisdennis replied:
No objections from me.

lorban commented on management/src/main/java/org/ehcache/management/registry/DefaultCollectorService.java:
```diff
@@ -84,7 +85,9 @@ public synchronized void start(ServiceProvider serviceProvider) {
     scheduledExecutorService = serviceProvider.getService(ExecutionService.class).getScheduledExecutor(configuration.getCollectorExecutorAlias());
 
-    cacheManager.registerListener(this);
+    if (cacheManager instanceof ObservableCacheManager) {
```
If this test evaluates to false, this is going to make the cache manager silently un-monitorable. There should at least be a warning about that.

cljohnso replied:
Hmmm ... looks like I dropped the `TODO` in one of my rebases.

This affects `DefaultCollectorService`, `DefaultSharedManagementService`, and `DefaultManagementRegistryService`.  Do we need the warning in each place?

mathieucarbou replied:
@cljohnso @lorban @chrisdennis @ljacomet : I have an alternative solution:

When I created the `CacheManagerProviderService`, the goal was to be able within the services to access a cache manager. Firstly because if the m&m needs (to listen on some events).

So perhaps instead of checking whether the CM is an instance of Observable, and silently discard a feature the user explicitly installed, why don't we change the `CacheManagerProviderService ` so that we have something like this:

```
CacheManager cacheManager = serviceProvider.getService(CacheManagerProviderService.class).getCacheManager();

ObservableCacheManager cacheManager = serviceProvider.getService(CacheManagerProviderService.class).getObservableCacheManager();
```
`getObservableCacheManager` would do a sanity check and throw an exception if the CM is not observable.

or...

`ObservableCacheManager cacheManager = serviceProvider.getService(CacheManagerProviderService.class).getCacheManager(ObservaleCacheManager.class);`

and the method `getCacheManager(Class<T extends CacheManager>)` would "unwrap" to the wanted class or throw.

The idea behind this is that a service gets what it needs, and if he can't then it crashes simply because it won't work anyway. I think it's not the responsibility of the service to check / know all the different potential flavors of a CM (because at the end there could be more than just the Observable flavor). The service just need to ask what it needs to work properly, and the init just need to fail if what is needed cannot be provided. It would be the responsibility of the developer to fix such error.

Management services are optional (theoretically because jsr107 depends on it at the moment, at least until #747 is fixed). So if somebody register such services within a CM that is not observable, it should fail at init (so when the cache manager provider would throw an exception), because the required feature won't work.

Let me know what you decided: I am interested in following the changes in the management module :-)

lorban commented on management/src/main/java/org/ehcache/management/registry/DefaultSharedManagementService.java:
```diff
@@ -53,39 +54,42 @@
-      @Override
-      public void cacheAdded(String alias, Cache<?, ?> cache) {
-      }
+    if (cacheManager instanceof ObservableCacheManager) {
```
Same as above: if this evaluates to false you silently have no monitoring.

ljacomet commented:
Can you get rid of the max memory commit? Something similar is now part of trunk

cljohnso commented:
I'll remove the max memory commit at my next rebase.

chrisdennis commented on build.gradle:
```diff
@@ -79,6 +79,10 @@ subprojects {
+  test {
+    jvmArgs "-Xmx256m"
+  }
+
```
Louis handled this in another commit so it should disappear on rebase I think.

cljohnso replied:
It will disappear.

chrisdennis commented on clustered/client/src/main/java/org/ehcache/clustered/config/ClusteringServiceConfiguration.java:
```diff
@@ -0,0 +1,68 @@
+    if (connectionUrl == null) {
+      throw new NullPointerException("connectionUrl");
+    }
+    this.alias = (alias == null ? "" : alias);
```
Why would we support a null alias?

cljohnso replied:
Don't know that we should.  Do we want to enforce a non-empty alias?

chrisdennis replied:
That depends... what does the alias mean?  Does it tie a cache to the clustering service that backs it, or does it identify the name of the entity to connect to?

chrisdennis commented on clustered/client/src/main/java/org/ehcache/clustered/config/ClusteringServiceConfiguration.java:
```diff
@@ -0,0 +1,68 @@
+    return alias;
+  }
+
+  public URI getConnectionUrl() {
```
Which is it?  URI or URL?

cljohnso replied:
From an XML point of view, it comes across as a URI.  From the programmatic specification, we could limit it to a URL ... we _are_ supposed to be able to connect to the thing, right?  Not only that, we could limit it to `http`/`https` URLs -- unless we want to designate a scheme of our own.

I have altered the extension schema to ensure the connection URL is a reasonably formed `http`/`https URL.  That will be included in the next commit.

chrisdennis replied:
We are supposed to connect to it... but generally I prefer sticking to URI until the last minute, since URLs have a nasty tendency to trigger DNS calls at the wrong moments.

cljohnso replied:
Yes ... for some reason, `equals` wants to resolve the host ... would have been better if that wasn't part of the `URL.equals` contract.

chrisdennis commented on clustered/client/src/main/java/org/ehcache/clustered/config/xml/ClusteredCacheConstants.java:
```diff
@@ -0,0 +1,37 @@
+  /**
+   * Name of schema resource file.
+   */
+  static final String XSD = "ehcache-clustered-ext.xsd";
```
Do we ever use the string constant directly?

cljohnso replied:
It was, at one time, used in a test -- that is no longer the case.  It is, however, referred to in javadoc in `ClusteringServiceConfigurationParser` and, assuming we continue the transaction model and use the same extension schema for `<cache>` extensions, will be referred to in the javadoc `CacheServiceConfigurationParser` implementation handling the `<cache>` extensions.

chrisdennis commented on clustered/client/src/main/java/org/ehcache/clustered/config/xml/ClusteringServiceConfigurationParser.java:
```diff
@@ -0,0 +1,104 @@
+            final Attr urlAttribute = ((Element)item).getAttributeNode("url");
+            final String urlValue = urlAttribute.getValue();
+            try {
+              connectionUri = URI.create(urlValue);
```
You should be calling new URI here as urlValue isn't a compile time constant.

cljohnso replied:
Doesn't change much but I see your point ... will change.

cljohnso replied:
BTW ... right now, the call path to this method makes no provisions for directly handling faulty input.  Should it?  I'm throwing a `RuntimeException` because that was how it was handled elsewhere.  As Ludovic mentioned elsewhere, `RuntimeException` isn't very "expressive".

chrisdennis commented on impl/src/main/java/org/ehcache/config/builders/CacheManagerBuilder.java:
```diff
@@ -68,37 +70,46 @@ public T build() {
 
   public <K, V> CacheManagerBuilder<T> withCache(String alias, CacheConfiguration<K, V> configuration) {
-    return new CacheManagerBuilder<T>(this, configBuilder.addCache(alias, configuration));
+    return this.newBuilder(configBuilder.addCache(alias, configuration));
```
Why all the 'this' references?

cljohnso replied:
For of habit from many years spent programming Java where `this` was a requirement ...

chrisdennis commented on impl/src/main/java/org/ehcache/config/builders/CacheManagerBuilder.java:
```diff
@@ -212,4 +230,52 @@ T cast(EhcacheManager ehcacheManager) {
+   *        {@link CacheManagerConfiguration} instances appearing in the configuration does
+   *        not build a {@code CacheManager} of the {@code specialization} type
+   */
+  public static <T extends CacheManager> T newSpecializedCacheManager(Class<T> specialization, Configuration config)
```
Does this need to be public?

cljohnso replied:
It's presently used from the `org.ehcache.clustered.ClusteredCacheManagerBuilder` class ...

cljohnso commented:
Rebased, made some review adjustments, and dropped the max memory commit.

In this update, I did **not** address `ClusteringServiceConfiguration` _alias_ handling or the `CacheManagerProvider` service changes suggested by @mathieucarbou -- which @chrisdennis has recommended handling in this PR.

cljohnso commented:
The latest commit addresses `ClusteringServiceConfiguration` _alias_ and the `CacheManagerProviderService` changes.

After discussion with @chrisdennis, `ObservableCacheManager` is renamed to `InternalCacheManager`.  The intent is that `InternalCacheManager` be the interface consumed internally (by services) and includes methods needed internally that we do not wish to expose on `CacheManager` -- such as those needed by the management layer.  `CacheManagerProviderService`  is changed to return an `InternalCacheManager` instead of `CacheManager` or one of the concrete implementations.  This was considered a better approach than introducing a casting-based approach.

The `ClusteringServiceConfiguration` _alias_ field was dropped -- it will be added in the future as needed (though, given other identification mechanisms, it's not likely to be necessary).

`ObservableCacheManager`/`InternalCacheManager` was relocated -- a package change (adding 'core' was previously missed).

cljohnso commented:
@chrisdennis @lorban @ljacomet 
We have yet to resolve the logging discussion anchored against `BaseCacheManager` ...

chrisdennis commented:
I'm happy with this.  @lorban @ljacomet I'd prefer if one of you guys gave this another pass and merged if you are happy too.

mathieucarbou commented:
:+1:
